tests/func/test_ignore.py

```diff
@@ -418,3 +418,27 @@ def test_run_dvcignored_dep(tmp_dir, dvc, run_copy):
     tmp_dir.gen({".dvcignore": "dir\n", "dir": {"foo": "foo"}})
     run_copy(os.path.join("dir", "foo"), "bar", name="copy-foo-to-bar")
     assert (tmp_dir / "bar").read_text() == "foo"
+
+
+def test_pull_ignore(tmp_dir, dvc, local_cloud):
+    from dvc.utils.fs import remove
+
+    tmp_dir.dvc_gen(
+        {
+            ".dvcignore": "data/processed/",
+            "data": {"foo": "foo", "processed": {"bar": "bar"}},
+        }
+    )
+    tmp_dir.add_remote(config=local_cloud.config)
+    dvc.add("data")
+    dvc.push()
+
+    foo_path = tmp_dir / "data" / "foo"
+    foo_path.unlink()
+    assert not foo_path.exists()
+
+    remove(tmp_dir / ".dvc/cache")
+    dvc.pull()
```

Comment:
If I change this to `dvc.checkout()` then it fails again with the same error about removing files without confirmation. I tried to fix this but got stuck, I'm not familiar enough with this codebase. I can share what I found while debugging. I'm actually not even sure if the error is expected or not. The issue gave me the impression that both `pull` and `checkout` should succeed, is that correct?

Reply:
Right, so basically it seems that `checkout` breaks down when the cache is deleted, which seems fair and maybe not worth trying to change? `fetch` updates the cache, I'm guessing `status -c` does too. And `pull` does a `fetch` before `checkout`.

Do you know if there are other ways to trigger this bug of `checkout` previously not ignoring files properly that don't involve deleting the cache?

Reply:
`status -c` doesn't pull down the cache but it will pull down the `.dir` index.  And it is used in generating a data tree object. 
https://github.com/iterative/dvc/blob/954de0cae44e31dfcf45630a2cc7e7f5acc7f00c/dvc/output.py#L772
I think the current problem is that,  in `dvc-data` the `diff` function didn't take `dvcignore` into account when the tree object is missing.

https://github.com/iterative/dvc-data/blob/010cdfb0527209eebbcec4a33c476cf4f532f963/src/dvc_data/checkout.py#L252

Reply:
To solve this completely solve this problem, some modifications to `dvc-data` is required.

Reply:
Would you like to solve it yourself or wait for someone else to do this? But I think it is different problem. We can merge this one as it already solved one part of the question.

Reply:
I don't know how to solve it myself, someone else should do this.